File: src/native/printer-plugin/registry/driver-registry.ts
import type { PrinterDriverType } from '../definitions'
import type { PrinterDriver } from '../internal/types'

export class DriverRegistry {
  private readonly factories = new Map<PrinterDriverType, () => PrinterDriver>()

  register(type: PrinterDriverType, factory: () => PrinterDriver): void {
    this.factories.set(type, factory)
  }

  resolve(type: PrinterDriverType): PrinterDriver {
    const factory = this.factories.get(type)
    if (!factory) {
      throw new Error(`Driver "${type}" chưa được đăng ký.`)
    }
    return factory()
  }
}

File: android/app/src/main/java/com/ndtcore/pos/printer/registry/DriverRegistry.kt
package com.ndtcore.pos.printer.registry

import com.ndtcore.pos.printer.driver.PrinterDriver

class DriverRegistry {
    private val factories = mutableMapOf<String, () -> PrinterDriver>()

    fun register(type: String, factory: () -> PrinterDriver) {
        factories[type] = factory
    }

    fun resolve(type: String): PrinterDriver {
        val factory = factories[type]
            ?: throw IllegalArgumentException("Driver \"$type\" chưa được đăng ký.")
        return factory()
    }
}
